content/en/blog/2023/otel-dotnet-performance-impacts.md

```diff
@@ -0,0 +1,288 @@
+linkTitle: OTel .NET Performance Impacts
+date: 2023-03-10
+author: '[MartinDotNet](https://github.com/martinjt) (Honeycomb)'
+---
+
+Contrary to
+[Betteridge’s Law of Tabloid Headlines](https://en.wikipedia.org/wiki/Betteridge%27s_law_of_headlines),
+the answer is “yes,” but context is important. I get this question so often that
+I thought it was time to get some stats on it.
+
+I’ve heard comments like:
+
+- “Instrumenting my code will make it slower”
+- “Activity Object use memory”
+- “If I instrument my code, it will cost me more as I’ll need bigger servers”
+- “OpenTelemetry is immature and isn’t memory efficient”
+- “OpenTelemetry causes performance issues”
+
+I can only assume that these are based on previous versions, or things like
+OpenTracing/OpenCensus (the heritage frameworks that were the feeders for
+OpenTelemetry). This is not true of the current state of OpenTelemetry, and in
+this post, I’ll provide evidence as to why this isn’t the case. My hope is that
+this evidence will stop the debate of performance impact so that teams can
+instead debate the value of instrumenting their code.
+
+TL;DR: if you’re optimizing at the level where OpenTelemetry and Activity usage
+would be a concern, you should probably be more worried about the .NET as a
+whole.
+
+## Activity vs. OpenTelemetry vs. OpenTelemetry Shim
+
+The first thing to note is developers just now getting to grips with the
+difference between using the `Activity` class in .NET and using the
+OpenTelemetry SDKs to process and export telemetry to a backend.
+
+Activity is the base of all instrumentation in .NET. It’s not tied to a third
+party, or even a particular OpenSource project like OpenTelemetry. It’s in the
+Base Class Library (BCL), which means it’s baked into .NET. You can instrument
+your code with this without deciding how you’re going to extract.
+
+OpenTelemetry is an open source SDK that takes generated `Activity` objects and
+sends them to a telemetry backend (whether it’s open source like Jaeger, or
+proprietary like Honeycomb or Lightstep).
+
+OpenTelemetry Shim is an SDK that was developed by the OpenTelemetry project to
+provide consistent naming across different languages. As .NET already had
+`Activity`, it was reused, but this meant that OpenTelemetry terminology wasn’t
+consistent over different languages. As such, Shim was created so engineers
+could use the same terminology across languages. Under the hood, Shim still
+creates `Activity` objects for Spans, and ActivitySource objects for Tracers.
+Therefore, there is a minor impact of using Shim for the wrapper object
+initialization.
+
+## Existing benchmarks
+
+For a while now, the OpenTelemetry .NET maintainers have had a BenchmarkDotNet
+project that tests the OpenTelemetry components. This is great for showing that
+the library doesn’t add any issues between releases. What it fails to answer is
+“what if I start with nothing?” Those tests are the inspiration for both tests I
```

Comment:
> What it fails to answer is
> “what if I start with nothing?” 

Could you elaborate which part it fails to answer? The benchmarks show the cost of producing an activity, and that'll be the minimum cost one has to pay when they enable otel in their app/library.